File: eidocell-ui/src/types/gallery.ts
export interface SampleOut {
  id: string
  filename: string
  path: string
  is_active: boolean
  class_id: string | null
  class_name: string | null
  class_color: string | null
  has_mask: boolean
  mask_channels: number[]
  n_channels: number
}

export interface SamplePage {
  items: SampleOut[]
  total: number
  offset: number
  limit: number
  session_has_any_mask: boolean
}

export interface FilterCondition {
  field: string
  operator: string
  value: string | number
  value2?: string | number | null
}

export interface SampleListParams {
  offset?: number
  limit?: number
  sort_by?: string
  sort_order?: string
  filters?: FilterCondition[]
}

export interface ClassCreate {
  name: string
  color?: string
}

export interface ClassUpdate {
  name?: string
  color?: string
}

export interface ClassOut {
  id: string
  name: string
  color: string
  sample_count: number
}

export interface BulkClassAssignment {
  sample_ids: string[]
  class_id: string | null
}

export interface SortableAttribute {
  name: string
  channel_index: number
  channel_name: string
  /** sort_by token, e.g. "attr:area:ch:0" */
  value: string
}

File: eidocell-backend/schemas/workspace/gallery.py
from pydantic import BaseModel


class SampleOut(BaseModel):
    id: str
    filename: str
    path: str
    is_active: bool
    class_id: str | None
    class_name: str | None
    class_color: str | None
    has_mask: bool
    mask_channels: list[int] = []
    n_channels: int = 1

    model_config = {"from_attributes": True}


class SamplePage(BaseModel):
    items: list[SampleOut]
    total: int
    offset: int
    limit: int
    session_has_any_mask: bool = False


class FilterCondition(BaseModel):
    field: str  # "filename", "class_name", or a mask attribute key
    operator: str  # "==", "!=", ">", ">=", "<", "<=", "between", "contains", "not_contains"
    value: str | float | int
    value2: str | float | int | None = None  # for "between"


class SampleListParams(BaseModel):
    offset: int = 0
    limit: int = 100
    sort_by: str = "filename"
    sort_order: str = "asc"  # "asc" or "desc"
    filters: list[FilterCondition] = []


class ClassCreate(BaseModel):
    name: str
    color: str = "#808080"


class ClassUpdate(BaseModel):
    name: str | None = None
    color: str | None = None


class ClassOut(BaseModel):
    id: str
    name: str
    color: str
    sample_count: int

    model_config = {"from_attributes": True}


class BulkClassAssignment(BaseModel):
    sample_ids: list[str]
    class_id: str | None  # None to unassign
